File: algorithm/str/kmp.cpp

#include <iostream>

using namespace std;

class Kmp
{
private:
    int *next;

public:
    Kmp(/* args */);
    int *GetNext(string p);
    int Find(string s, string p);
    ~Kmp();
};

Kmp::Kmp(/* args */)
{
}

Kmp::~Kmp()
{
    // free(this->next);
}

int *Kmp::GetNext(string p)
{
    int n = p.length();
    this->next = new int[n];
    this->next[0] = -1;

    for (int i = 0, j = -1; i < p.length() - 1;)
    {
        if (j == -1 || p[i] == p[j])
        {
            i++;
            j++;
            this->next[i] = j;
        }
        else
        {
            i++;
            j = this->next[j];
        }
    }

    // for (int i = 0; i < n; i++)
    // {
    //     cout << this->next[i] << " ";
    // }
    // cout << endl;
    return this->next;
}

int Kmp::Find(string s, string p)
{
    int pos = -1;
    int i = 0, j = 0;
    if (s.length() == 0 || p.length() == 0 || p.length() > s.length())
    {
        return -1;
    }

    this->next = this->GetNext(p);

    while (j < p.length())
    {
        if (s[i] == p[j])
        {
            i++;
            j++;
            if (j == p.length()){
                pos = i - j;
                return pos;
            }
        }
        else
        {
            j = this->next[j];
            if (j == -1)
            {
                j++;
                i++;
            }
            if (s.length() - i < p.length() - j){
                return -1;
            }
        }
    }

    return pos;
}

int main(int argc, char const *argv[])
{
    Kmp k;
    cout << k.Find("abcdaaaabcabcdebbb", "abcabcde") << endl;
    return 0;
}

File: algorithm/str/kmp.py

class kmp:
    def __call__(self, s, p):

        if not s or not p or (len(p) > len(s)): # 边界：都为空字符串，或模式串为空，或模式串比源串长
            return -1
        
        next_list = self.get_next(p)
        pos = -1

        i = j = 0
        while j < len(p):
            if s[i] == p[j]:
                i += 1
                j += 1
                if j == len(p):
                    pos = i - j
                    break
            else:
                j = next_list[j]
                if j == -1:
                    i += 1
                    j = 0
                if len(s) - i < len(p) - j: # 剩余子串不够匹配
                    break
        return pos


    def get_next(self, p):
        '''
            'abcabcde' -> [-1, 0, 0, 0, 1, 2, 3, 0]
            next列表存放的是，模式串第i位从头开始有多少位是相同的。
            头部增加1个'-1'值，便于操作。
        '''
        i = 0
        j = -1
        next_list = [-1]

        while i < len(p) - 1:
            if j == -1 or p[i] == p[j]:
                i += 1
                j += 1
                next_list.append(j)
            else:
                j = next_list[j]
        return next_list

if __name__ == "__main__":
    s = 'abcdaaaabcabcdebbb'
    k = kmp()
    print(k(s, 'abcabcde'))